services/code_hosting/git_hub_service.rb

```diff
@@ -1,131 +1,409 @@
+
+      attr_accessor :temporary_storage_path
+
+      attr_writer :cache
+
+      attr_accessor :git_action_queue
+
+      def cache
+        @cache ||= {}
+        return @cache
+      end
+
+      # Client for GitHub-related operations, uses an in-memory cache
+      # to fast things.
+      # @param [String] api_token
+      def client(api_token)
+        @client_cache ||= {}
+        return @client_cache[api_token] unless @client_cache[api_token].nil?
+        client = Octokit::Client.new(access_token: api_token)
+        @client_cache[api_token]
+        return client
+      end
+
+      def temp_path(root_path = Dir.tmpdir)
+        temporary_path = File.join(root_path, ".fastlane")
+        FileUtils.mkdir_p(temporary_path) unless File.directory?(temporary_path)
+        return temporary_path
+      end
+
+      def temporary_git_storage(root_path = Dir.tmpdir)
+        temp_storage = File.join(root_path, ".tmp")
+        @temporary_git_storage ||= temp_storage
+        FileUtils.mkdir_p(@temporary_git_storage) unless File.directory?(@temporary_git_storage)
+        return @temporary_git_storage
+      end
+
+      # This method shallow clones the repo using the provider credential and looks for a valid _fastlane_
+      # configuration, parses the Fastfile (if found), and returns the lanes in the following way:
+      #  {
+      #    "no_platform": {
+      #      "": {
+      #        "description": [
+      #
+      #        ],
+      #        "actions": [
+      #          {
+      #            "action": "default_platform",
+      #            "parameters": "ios"
+      #          }
+      #        ]
+      #      }
+      #    },
+      #    "ios": {
+      #      "test": {
+      #        "description": [
+      #          "Description of what the lane does"
+      #        ],
+      #        "actions": [
+      #          {
+      #            "action": "gym",
+      #            "parameters": {
+      #              "skip_package_ipa": false,
+      #              "clean": true,
+      #              "project": "../fastlane-ci-demoapp.xcodeproj",
+      #              "scheme": "fastlane-ci-demoapp"
+      #            }
+      #          }
+      #        ],
+      #        "private": false
+      #      }
+      #    }
+      #  }
+      #
+      # @param repo_url [String]
+      # @param branch [String]
+      # @param path [String]
+      # @param provider_credential [GithubProviderCredential]
+      # @return [Hash<String, Hash>] being the key the name of the platform (:ios, :android, :no_platform) and the Hash the underlying lane (name, actions).
+      def peek_fastfile_configuration(repo_url: nil, branch: nil, provider_credential: nil, path: self.temp_path, cache: true)
+        repo = repo_from_url(repo_url)
+        return self.cache[[repo, branch].join("/")] if cache && self.cache && !self.cache[[repo, branch].join("/")].nil? && self.cache[[repo, branch].join("/")].kind_of?(Hash)
```

Comment:
this is a really complicated line, let's see if there's a way we can make this more imperative so people can understand each check you're making.

Reply:
I'll split this in several statements so is more explicit for people to understand what is really doing.